allennlp/training/trainer.py

```diff
@@ -820,6 +829,10 @@ def train(self) -> Dict[str, Any]:
 
             epochs_trained += 1
 
+        metrics['peak_cpu_memory_usage_MB'] = max_cpu_usage
```

Comment:
I'd move both modifications to this method into a single block like the `if is_best_so_far` block:

```python
if 'cpu_memory' in metrics:
    metrics['peak_cpu_memory'] = max(metrics.get('peak_cpu_memory', 0), metrics['cpu_memory'])
if 'gpu_memory' in metrics:
    metrics['peak_gpu_memory'] = max(metrics.get('peak_gpu_memory', 0), metrics['gpu_memory'])
```

You'll need to modify the second part of that a bit for multiple GPUs, but that's a lot simpler than the two separate blocks of changes that you have here.

Reply:
`2018-12-05T20:48:20.143898512Z	  File "/stage/allennlp/allennlp/commands/train.py", line 351, in train_model
2018-12-05T20:48:20.143901172Z	    metrics = trainer.train()
2018-12-05T20:48:20.143903453Z	  File "/stage/allennlp/allennlp/training/trainer.py", line 792, in train
2018-12-05T20:48:20.143906055Z	    self._metrics_to_tensorboard(epoch, train_metrics, val_metrics=val_metrics)
2018-12-05T20:48:20.143908459Z	  File "/stage/allennlp/allennlp/training/trainer.py", line 660, in _metrics_to_tensorboard
2018-12-05T20:48:20.143911078Z	    self._tensorboard.add_train_scalar(name, train_metric, epoch)
2018-12-05T20:48:20.143913417Z	  File "/stage/allennlp/allennlp/training/trainer.py", line 126, in add_train_scalar
2018-12-05T20:48:20.14391598Z	    self._train_log.add_scalar(name, self._item(value), global_step)
2018-12-05T20:48:20.143918549Z	  File "/usr/local/lib/python3.6/site-packages/tensorboardX/writer.py", line 272, in add_scalar
2018-12-05T20:48:20.143921039Z	    self.file_writer.add_summary(scalar(tag, scalar_value), global_step)
2018-12-05T20:48:20.14392354Z	  File "/usr/local/lib/python3.6/site-packages/tensorboardX/summary.py", line 87, in scalar
2018-12-05T20:48:20.143926035Z	    assert(scalar.squeeze().ndim == 0), 'scalar should be 0D'
2018-12-05T20:48:20.143928559Z	AttributeError: 'NoneType' object has no attribute 'squeeze'`

Reply:
I got an error like this after putting `gpu_memory` and `cpu_memory` directly in `train_metrics` returned from `_train_epoch`. Should I keep it in the `train_metrics`?

Reply:
Are you adding `Nones` anywhere?  I'm not sure why you are seeing a `NoneType` error.

Reply:
Maybe there's no GPU usage, because you have no GPU, and you shouldn't be adding `gpu_usage` to the metrics in that case?

Reply:
`cpu_memory` is returned by `peak_memory_mb()` and `gpu_usage` will be a empty list if there's no gpu, and I don't add `gpu_usage` into metrics if it's empty. So I don't think any of them will become None after returned. I was thinking if the tensorboard can recognize them??